lib/auth/join.go

```diff
@@ -114,6 +115,92 @@ func setRemoteAddrFromContext(ctx context.Context, req *types.RegisterUsingToken
 	return nil
 }
 
+// handleJoinFailure logs and audits the failure of a join. It is intentionally
+// designed to handle potential nullness of the input parameters.
+func (a *Server) handleJoinFailure(
+	origErr error,
+	pt types.ProvisionToken,
+	attributeSource joinAttributeSourcer,
+	req *types.RegisterUsingTokenRequest,
+) {
+	fields := logrus.Fields{
+		"role": "unknown",
+	}
+	if req != nil {
+		fields["role"] = req.Role
+	}
+
+	var attributesProto *apievents.Struct
+	if attributeSource != nil {
+		var err error
+		attributes, err := attributeSource.JoinAuditAttributes()
+		if err != nil {
+			log.WithError(err).Warn("Unable to fetch join attributes from join method")
+		}
+		fields["attributes"] = attributes
+		attributesProto, err = apievents.EncodeMap(attributes)
+		if err != nil {
+			log.WithError(err).Warn("Unable to encode join attributes for audit event")
+		}
+	}
+
+	if pt != nil {
+		fields["join_method"] = string(pt.GetJoinMethod())
+		fields["token_name"] = pt.GetSafeName()
+	}
+	log.WithError(origErr).WithFields(fields).Error("Failure to join cluster occurred")
+
+	var evt apievents.AuditEvent
+	status := apievents.Status{
+		Success: false,
+		Error:   origErr.Error(),
+	}
+	if req != nil && req.Role == types.RoleBot {
+		botJoinEvent := &apievents.BotJoin{
+			Metadata: apievents.Metadata{
+				Type: events.BotJoinEvent,
+				Code: events.BotJoinFailureCode,
+			},
+			Status:     status,
+			BotName:    "unknown",
+			Method:     "unknown",
```

Comment:
is it our convention to add `"unknown"` in all of these fields instead of leaving them empty?

Reply:
Hm, I'm unsure as well. To me, `"unknown"` throws a much bigger red flag than empty but I could be wrong. If we've already implemented other unknowns, then perhaps we shouldn't change it here until we have a better idea/plan. If it's a UI thing, we could bring it up at the UX/UI open hours to see if they have a better solution?

Reply:
It looks like I did something similar in various desktop events e.g.

https://github.com/gravitational/teleport/blob/39882ea520705b8a6ff8114a5b7653117a1a81de/lib/srv/desktop/audit.go#L267

It could make sense to standardize this along the lines of:

- `""`: consider this a bug
- `"unknown"`: typically events of this kind would have this field, but it could not be known at the time of event emission

The idea being that we want to be able to clearly distinguish between data loss due to bugs (which are high priority fixes, due to the security-relevance of audit events) vs data which we explicitly omit by necessity.

However I notice that e.g `BotName` is `"omitempty"`

https://github.com/gravitational/teleport/blob/39882ea520705b8a6ff8114a5b7653117a1a81de/api/proto/teleport/legacy/types/events/events.proto#L3740-L3741

so this particular case might be as @avatus suggested, just worthy of a client side display fix.

Reply:
> It could make sense to standardize this along the lines of:
> 
> - "": consider this a bug
> - "unknown": typically events of this kind would have this field, but it could not be known at the time of event emission

Imho this is a good direction to head in - audit logs contain important information and when this information is missing, we need to act. Distinguishing between audit logs that have been emitted in a place where a certain field is not yet known, and those that are missing the information because of a bug or because they predate the field having been added is useful.

> However I notice that e.g BotName is "omitempty"

I'm not too sure this has a significant impact, at the time when I added these events, I wasn't too sure whether to use omitempty or not.

Reply:
After a chat with Edoardo, I'm going to remove these "unknown" default values for now.

Reply:
tl;dr of your chat out of curiousity?